### ISSUE TITLE
Provisioning devices using Enrollment groups

### ISSUE THREAD

**Issue by rchakra3** [Azure/azure-iot-sdk-python#226]
- **OS and version used:** Ubuntu 16.04

- **Python runtime used:** python 3.7.0

- **SDK version used:** azure-iot-provisioning-device-client v1.4.3

# Description of the issue:
I'm trying to enroll a device using the group enrollment option. 

I have a verified CA cert in the service and an enrollment group that's configured with this cert. How do I enroll a new device with a certificate signed by that CA? The [docs](https://docs.microsoft.com/en-us/azure/iot-dps/quick-enroll-device-x509-java#modifications-to-enroll-a-single-x509-device) have examples for Java but nothing for python. Is this possible using the python SDK? 

Also, probably related - I tried using the device provisioning sample code and seems to be looking for a cert based on the output pasted below. Where exactly is it looking for the cert? 

# Code sample exhibiting the issue:
https://github.com/Azure/azure-iot-sdk-python/blob/master/provisioning_device_client/samples/provisioning_device_client_sample.py

# Console log of the issue:
When I try using the provisioning sample I get the following errors:
```
{"operationId":"4.87f7851aaaa18ed1.9123edc4-4058-4f78-a320-cf6cac5cff13","status":"failed","registrationState":{"registrationId":"riot-device-cert","createdDateTimeUtc":"2018-10-29T23:04:14.4198475Z","status":"failed","errorCode":401002,"errorMessage":"CA certificate not found","lastUpdatedDateTimeUtc":"2018-10-29T23:04:14.6002502Z","etag":"\u00222c00feaa-0000-0000-0000-5bd791ee0000\u0022"}}

Error: Time:Mon Oct 29 23:04:17 2018 File:/usr/sdk/src/c/provisioning_client/src/prov_device_ll_client.c Func:prov_transport_process_json_reply Line:406 Provisioning Failure: OperationId: 4.87f7851aaaa18ed1.9123edc4-4058-4f78-a320-cf6cac5cff13 - Date: 2018-10-29T23:04:14.6002502Z - Msg: CA certificate not found
```

**Comment by SyedOsman33** [comment]
> * **OS and version used:** Ubuntu 16.04
> * **Python runtime used:** python 3.7.0
> * **SDK version used:** azure-iot-provisioning-device-client v1.4.3
> 
> # Description of the issue:
> I'm trying to enroll a device using the group enrollment option.
> 
> I have a verified CA cert in the service and an enrollment group that's configured with this cert. How do I enroll a new device with a certificate signed by that CA? The [docs](https://docs.microsoft.com/en-us/azure/iot-dps/quick-enroll-device-x509-java#modifications-to-enroll-a-single-x509-device) have examples for Java but nothing for python. Is this possible using the python SDK?
> 
> Also, probably related - I tried using the device provisioning sample code and seems to be looking for a cert based on the output pasted below. Where exactly is it looking for the cert?
> 
> # Code sample exhibiting the issue:
> https://github.com/Azure/azure-iot-sdk-python/blob/master/provisioning_device_client/samples/provisioning_device_client_sample.py
> 
> # Console log of the issue:
> When I try using the provisioning sample I get the following errors:
> 
> ```
> {"operationId":"4.87f7851aaaa18ed1.9123edc4-4058-4f78-a320-cf6cac5cff13","status":"failed","registrationState":{"registrationId":"riot-device-cert","createdDateTimeUtc":"2018-10-29T23:04:14.4198475Z","status":"failed","errorCode":401002,"errorMessage":"CA certificate not found","lastUpdatedDateTimeUtc":"2018-10-29T23:04:14.6002502Z","etag":"\u00222c00feaa-0000-0000-0000-5bd791ee0000\u0022"}}
> 
> Error: Time:Mon Oct 29 23:04:17 2018 File:/usr/sdk/src/c/provisioning_client/src/prov_device_ll_client.c Func:prov_transport_process_json_reply Line:406 Provisioning Failure: OperationId: 4.87f7851aaaa18ed1.9123edc4-4058-4f78-a320-cf6cac5cff13 - Date: 2018-10-29T23:04:14.6002502Z - Msg: CA certificate not found
> ```

I am Facing the exact same issue, the sample in SDK only works for Individual Enrollments but in case of group enrollments there is no sample code provided and provisioning_device_client_sample.py do not work for Enrollment Groups

**Comment by anhashia** [comment]
Device provisioning sample currently uses a **sample** certificate for demonstration purpose **only**.

This sample certificate you get is mentioned in following location. 
https://github.com/Azure/azure-iot-sdk-python/blob/master/provisioning_device_client/doc/using_group_enrollment_dice.md
**dice_device_enrollment** is the **tool** you need to run 

Sample cert chain is like this
riot-root->riot-signer-core->riot-device-cert

Here leaf (device certificate) is : **riot-device-cert**
Intermediate CA  is : **riot-signer-core**
Root CA is : **riot-root**

There are 2 scenarios here
1.)  **Individual enrollment** 
In this scenario you will use the individual certificate generated from the tool (**option i**)  and upload the public certificate for riot-device-cert in DPS -> Manage enrollment blade (Azure Portal).
You registration id will be **riot-device-cert** once upload of PEM is successful.

**Option i** in the tool will give you 2 certificates : riot-device-cert and riot-signer-core.

After that you will run the sample
https://github.com/Azure/azure-iot-sdk-python/blob/master/provisioning_device_client/samples/provisioning_device_client_sample.py


2.) **Group enrollment**
Here you will need to run dice_device_enrollment tool with **option g**.
This will give you you riot-root certificate. You will have to do Proof of Possession on Root CA using verification code. The sample will prompt you to enter verification code.
After that you will create a Group Enrollment entry in Portal  and set Intermediate Certificate as **riot-signer-core**. This you will get from **option i** in the tool.

If running on Windows 10, make sure riot-device-cert is trusted so that during the TLS handshake client presets  riot-signer-core->riot-device-cert (chain) to the DPS service for authentication to succeed.

After that you will run the sample in same way you ran it for Individual enrollment
https://github.com/Azure/azure-iot-sdk-python/blob/master/provisioning_device_client/samples/provisioning_device_client_sample.py

This should result in Group Enrollment of the device with device id as riot-device-cert . There is no separate sample for Group Enrollment. It is the same sample.  Difference is on DPS service side on how to set the registration. 

As far as sample **not** providing a option to use your **own** certificates just like C# and Java SDK, it is covered under https://github.com/Azure/azure-iot-sdk-c/issues/635.

**Comment by pierreca** [comment]
I believe this has been addressed by @anhashia. Issue inactive for over 2 months. closing.

**Comment by az-iot-builder-01** [comment]
@rchakra3, @SyedOsman33, @anhashia, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by martgut** [comment]
Although this issue is closed, I'm blocked using the Azure IoT SDK for Python to accept devices with X.509 certificates, issued by our **proprietary** PKI. The enrollment groups is setup and prepared, but I'm failing to implement a Python simulator using a specific certificate/key, which has been issued by our PKI.

I'm following these instructions, but I cannot specify the certificate and key:
https://docs.microsoft.com/en-us/azure/iot-dps/quick-create-simulated-device-x509-python

What I'm actually missing is what I can do in Java:

`SecurityProvider securityProviderX509 = new SecurityProviderX509Cert(leafPublicPem, leafPrivateKey, signerCertificates);
provisioningDeviceClient = ProvisioningDeviceClient.create(globalEndpoint, idScope, PROVISIONING_DEVICE_CLIENT_TRANSPORT_PROTOCOL,
                                                                       securityProviderX509);`

As mentioned in other issues, the Python SDK is using a hardcoded certificate (riot-device-cert), which does not really help when using this in production scenarios. Help is greatly appreciated!

This is related to: #189, https://github.com/Azure/azure-iot-sdk-c/issues/635

**Comment by pierreca** [comment]
@martgut this version of the SDK is plagued by its dependency on the C SDK.
I believe this explains how you'd want to do things:
https://github.com/Azure/azure-iot-sdk-python/issues/254#issuecomment-483371875

i'm fully aware this is not a good experience. we're working on it. For now, unless you're willing to invest in developing a custom HSM module in C and recompile the SDK, there's unfortunately no good option for python developers.

**Comment by martgut** [comment]
@pierreca this is unfortunate, nevertheless really many thanks for the pointer!
Short term I will investigate the HSM option, whereby my preference would be a Python native solution like: 'azure-iot-sdk-python-preview'. Support of X.509 authentication for Python is a must for running our DevOps related tests.

**Comment by JMayrbaeurl** [comment]
@pierreca Any ETA on having this available in the SDK?

**Comment by pierreca** [comment]
related to #272 (and answered there)
TL;DR: fixed in v2
